### ISSUE TITLE
Regression: cannot run as non-root user with PSP in 0.78

### ISSUE THREAD

**Issue by Pluies** [emissary-ingress/emissary#1841]
**Describe the bug**

Hello!

Since upgrading to 0.78, we get this error message when starting Ambassador:
```
2019-09-18 16:12:42 diagd 0.78.0 [P120TAmbassadorEventWatcher] ERROR: could not reconfigure: [Errno 1] Operation not permitted: 'envoy'
2019-09-18 16:12:42 diagd 0.78.0 [P120TAmbassadorEventWatcher] ERROR: [Errno 1] Operation not permitted: 'envoy'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/ambassador-0.0.0.dev0-py3.6.egg/ambassador_diag/diagd.py", line 648, in run
    self.load_config_watt(rqueue, url)
  File "/usr/lib/python3.6/site-packages/ambassador-0.0.0.dev0-py3.6.egg/ambassador_diag/diagd.py", line 823, in load_config_watt
    self._load_ir(rqueue, aconf, fetcher, scc, snapshot)
  File "/usr/lib/python3.6/site-packages/ambassador-0.0.0.dev0-py3.6.egg/ambassador_diag/diagd.py", line 842, in _load_ir
    if not self.validate_envoy_config(config=ads_config, retries=self.app.validation_retries):
  File "/usr/lib/python3.6/site-packages/ambassador-0.0.0.dev0-py3.6.egg/ambassador_diag/diagd.py", line 1126, in validate_envoy_config
    odict['output'] = subprocess.check_output(command, stderr=subprocess.STDOUT, timeout=timeout)
  File "/usr/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.6/subprocess.py", line 423, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 1] Operation not permitted: 'envoy'
```

This leads to Envoy not starting, the pod being killed by failing its livenessprobe, and eventually a CrashLoopBackoff.

Looking into [the source](https://github.com/datawire/ambassador/blob/4a1ce2aaf854246a85acbd8fe3d02f1a808ddc9d/ambassador/ambassador_diag/diagd.py#L1122), this seems to point to the `envoy` binary not being executable by diagd, so I exec'd into the pod, which confirmed this:
```bash
[17:06:27]~/ λ kubectl exec -it ambassador-dev-5846c7d487-9kwn7 sh -cambassador
/ambassador $ ls
ambassador-config       demo-services           grab-snapshots.py       kubewatch.py            snapshots
ambassador-demo-config  entrypoint.sh           kick_ads.sh             post_update.py          watch_hook.py
ambex.pid               envoy                   kubestatus              post_watt.sh            watt
/ambassador $ envoy
sh: envoy: Operation not permitted
/ambassador $ which envoy
/usr/local/bin/envoy
/ambassador $ whoami
whoami: unknown uid 8888
/ambassador $
/ambassador $ /usr/local/bin/envoy
sh: /usr/local/bin/envoy: Operation not permitted
/ambassador $
```

I believe this is a combination of the new feature introduced by https://github.com/datawire/ambassador/pull/1720, which adds the capability onto this executable (not sure if that's the right way to put it, but you get it :) ) and our use of a PodSecurityPolicy that forces the pod to drop all capabilities:

```bash
[17:35:21]~/ λ kubectl get psp restricted
NAME         PRIV    CAPS   SELINUX    RUNASUSER          FSGROUP     SUPGROUP    READONLYROOTFS   VOLUMES
restricted   false          RunAsAny   MustRunAsNonRoot   MustRunAs   MustRunAs   false            configMap,downwardAPI,emptyDir,persistentVolumeClaim,secret,projected
```

**To Reproduce**

1. Set up a limited PodSecurityPolicy with RequiredDropCapabilities ([see docs](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#capabilities))
2. Deploy Ambassador's base Helm chart, that uses user 8888 by default
3. Pods won't start

**Expected behavior**

Ambassador should start normally as non-root user without needing additional capabilities when low ports are not needed.

**Versions (please complete the following information):**
 - Ambassador: 0.78.0
 - Kubernetes environment: EKS
 - Version: 1.14

**Additional context**

Version 0.77 is unaffected.

cc @swalberg @kflynn

**Comment by swalberg** [comment]
Can I see that PSP in yaml form so I can try it out for myself?

**Comment by Pluies** [comment]
Sure thing! Here it is:
```yaml
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  creationTimestamp: "2019-09-17T09:04:23Z"
  name: restricted
  resourceVersion: "1774"
  selfLink: /apis/extensions/v1beta1/podsecuritypolicies/restricted
  uid: 24a17fd8-d92a-11e9-936f-0a8b12d97aa0
spec:
  allowPrivilegeEscalation: false
  fsGroup:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
  requiredDropCapabilities:
  - ALL
  runAsUser:
    rule: MustRunAsNonRoot
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
  volumes:
  - configMap
  - downwardAPI
  - emptyDir
  - persistentVolumeClaim
  - secret
  - projected
```
We've lifted it up pretty much wholesale from the [restricted-psp.yaml](https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/policy/restricted-psp.yaml) example in [the official docs](https://kubernetes.io/docs/concepts/policy/pod-security-policy/).

**Comment by swalberg** [comment]
I wasn't able to figure out the pod security policies because I don't have that admission controller anywhere. But if I explicitly drop the capability in the pod spec, I get the same, e.g.

```
        # in the ambassador container spec
        securityContext:
          capabilities:
            drop: ["NET_BIND_SERVICE"]
```

```
2019-09-18 19:44:17 diagd 0.78.0 [P116TAmbassadorEventWatcher] ERROR: could not reconfigure: [Errno 1] Operation not permitted: 'envoy'
2019-09-18 19:44:17 diagd 0.78.0 [P116TAmbassadorEventWatcher] ERROR: [Errno 1] Operation not permitted: 'envoy'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/ambassador-0.0.0.dev0-py3.6.egg/ambassador_diag/diagd.py", line 648, in run
    self.load_config_watt(rqueue, url)
```

This feels like a product level decision -- should Ambassador be able to run in an environment that `CAP_NET_BIND_SERVICE` has been explicitly disallowed or should it be able to bind to a low port as a non root user? 

It would be nice to have the best of both worlds though, I'm open to suggestions.

**Comment by Pluies** [comment]
A potential avenue to allow both would be to have two `envoy` binaries – one with the `setcap` applied, and the other one without?

**Comment by Pluies** [comment]
Would be good to have opinions on the preferred way forward on this issue from project maintainers – @kflynn / @richarddli ? 🙂

**Comment by kflynn** [comment]
@Pluies We're actually digging into this a bit to see how we might be able to let this work for everyone. More soon. :)

**Comment by swalberg** [comment]
Currently looking into options. @Pluies, can you please get me the output of `grep Cap /proc/1/status` when inside an ambassador container? Thanks.

**Comment by Pluies** [comment]
@swalberg sure thing!

Here is the output in a Ambassador 0.77 pod running with the completely restricted PSP:
```
bash-4.4$ grep Cap /proc/1/status
CapInh:	0000000000000000
CapPrm:	0000000000000000
CapEff:	0000000000000000
CapBnd:	0000000000000000
CapAmb:	0000000000000000
```

We also upgraded to Ambassador 0.83 on some clusters due to the CVE. This required us to remove the rule to drop all Capacities from our PSP, and the output of this command in these pods is:

```
bash-5.0$ grep Cap /proc/1/status
CapInh:	00000000a80425fb
CapPrm:	0000000000000000
CapEff:	0000000000000000
CapBnd:	00000000a80425fb
CapAmb:	0000000000000000
```

**Comment by swalberg** [comment]
@khussey Are you having the same problem as above? The behaviour should have been reverted some time ago. #2050 is a potential fix to let you bind to low ports as a non root user.

**Comment by khussey** [comment]
@swalberg no, I just noticed that #2050 purports to fix it and figure either both the issue and PR ought to be open or both should be closed (I opted for the former). /cc @kflynn

**Comment by akevdmeer** [comment]
This would still be nice and it looks like the job is already half done. The capabilities_wrapper was merged in #1977 and remains, but the entrypoint can't yet make use of it as proposed in the now quite ancient #2050.

**Comment by Pluies** [comment]
PSPs are now deprecated so this issue is moot - closing.
